src/bin/tikv-ctl.rs

```diff
@@ -21,25 +21,415 @@ extern crate clap;
+        }
+    }
+
+    fn dump_raft_log(&self, region: u64, index: u64) {
+        let idx_key = keys::raft_log_key(region, index);
+        println!("idx_key: {}", escape(&idx_key));
+        println!("region: {}", region);
+        println!("log index: {}", index);
+
+        let mut entry = self.get_raft_log(region, index);
+        let data = entry.take_data();
+        println!("entry {:?}", entry);
+        println!("msg len: {}", data.len());
+
+        let mut msg = RaftCmdRequest::new();
+        msg.merge_from_bytes(&data).unwrap();
+        println!("{:?}", msg);
+    }
+
+    fn dump_mvccs_infos(
+        &self,
+        from: Vec<u8>,
+        to: Option<Vec<u8>>,
+        limit: Option<u64>,
+        cfs: Vec<&str>,
+        start_ts: Option<u64>,
+        commit_ts: Option<u64>,
+    ) {
+        let to = to.unwrap_or_default();
+        let limit = limit.unwrap_or_default();
+        if to.is_empty() && limit == 0 {
+            eprintln!(r#"please pass "to" or "limit""#);
+            process::exit(-1);
+        }
+        if !to.is_empty() && to < from {
+            eprintln!("The region's from pos must greater than the to pos.");
+            process::exit(-1);
+        }
+        let scan_future = self.get_mvcc_infos(from, to, limit).for_each(
+            move |(key, mut mvcc)| {
+                println!("key: {}", escape(&key));
+                if cfs.contains(&CF_LOCK) && mvcc.has_lock() {
+                    let mut lock_info = mvcc.take_lock();
+                    if start_ts.map_or(true, |ts| lock_info.get_lock_version() == ts) {
+                        // FIXME: "lock type" is lost in kvproto.
+                        let pk = escape(lock_info.get_primary_lock()).into_bytes();
+                        let k = escape(lock_info.get_key()).into_bytes();
+                        lock_info.set_primary_lock(pk);
+                        lock_info.set_key(k);
+                        println!("\tlock cf value: {}", print_to_string(&lock_info));
+                    }
+                }
+                if cfs.contains(&CF_DEFAULT) {
+                    for mut value_info in mvcc.take_values().into_iter() {
+                        if commit_ts.map_or(true, |ts| value_info.get_ts() == ts) {
+                            let value = escape(value_info.get_value()).into_bytes();
+                            value_info.set_value(value);
+                            println!("\tdefault cf value: {}", print_to_string(&value_info));
+                        }
+                    }
+                }
+                if cfs.contains(&CF_WRITE) {
+                    for write_info in mvcc.take_writes().into_iter() {
+                        if start_ts.map_or(true, |ts| write_info.get_start_ts() == ts) &&
+                            commit_ts.map_or(true, |ts| write_info.get_commit_ts() == ts)
+                        {
+                            // FIXME: short_value is lost in kvproto.
```

Comment:
Why is it a problem?

Reply:
The pb `WriteInfo` has not `short_value` field, so we can't show that.